externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala

```diff
@@ -64,11 +70,43 @@ case class SparkSQLEngine(spark: SparkSession) extends Serverable("SparkSQLEngin
       assert(currentEngine.isDefined)
       currentEngine.get.stop()
     })
+
+    startLifetimeChecker(() => {
+      assert(currentEngine.isDefined)
+      currentEngine.get.stop()
+    })
+
+  }
+
+  override def stop(): Unit = synchronized {
+    super.stop()
+
+    shutdown = true
+    val shutdownTimeout: Long = conf.get(ENGINE_EXEC_POOL_SHUTDOWN_TIMEOUT)
+    ThreadUtils.shutdown(lifetimeChecker, Duration(shutdownTimeout, TimeUnit.MILLISECONDS))
   }
 
   override protected def stopServer(): Unit = {
     countDownLatch.countDown()
   }
+
+  private[kyuubi] def startLifetimeChecker(stop: () => Unit): Unit = {
+    val interval = conf.get(ENGINE_CHECK_INTERVAL)
+    val maxLifetime = conf.get(ENGINE_SPARK_MAX_LIFETIME)
+    if (maxLifetime > 0) {
+      val checkTask = new Runnable {
+        override def run(): Unit = {
+          val lifetime: Long = System.currentTimeMillis() - getStartTime
+          val openSessionCount: Int = backendService.sessionManager.getOpenSessionCount
+          if (!shutdown && lifetime > maxLifetime && openSessionCount <= 0) {
+            info(s"Spark engine has been running for more than $maxLifetime ms, terminating")
+            stop()
```

Comment:
We need deregister the engineServiceDiscovery at first.

And then if the current open session count is zero, we can stop the engine, otherwise, we need wait.

```
if (!shutdown && lifetime > maxLifetime) {
    
   if (unDeregistered) {
      info (".... deregistering ...")
      frontendServices.flatMap(_.discoveryService).map {
        case engineServiceDiscovery: EngineServiceDiscovery =>
          engineServiceDiscovery.stop()
      }
   }

  if (openSessionCount <= 0) {
            info(s"Spark engine has been running for more than $maxLifetime ms, terminating")
            stop()
  }
}


```